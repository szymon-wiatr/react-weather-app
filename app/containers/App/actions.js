import {
  LOAD_WEATHER,
  LOAD_WEATHER_SUCCESS,
  LOAD_WEATHER_ERROR,
} from './constants';

/**
 * Load the weather, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_WEATHER
 */
export function loadWeather() {
  return {
    type: LOAD_WEATHER,
  };
}

/**
 * Dispatched when the weather are loaded by the request saga
 *
 * @param  {array} weather The weather data
 * @param  {string} cityname The current cityname
 *
 * @return {object} An action object with a type of LOAD_WEATHER_SUCCESS passing the weather
 */
export function weatherLoaded(weather, cityname) {
  return {
    type: LOAD_WEATHER_SUCCESS,
    weather,
    cityname,
  };
}

/**
 * Dispatched when loading the wheater fails
 *
 * @param  {object} error The error
 *
 * @return {object} An action object with a type of LOAD_WEATHER_ERROR passing the error
 */
export function weatherLoadingError(error) {
  return {
    type: LOAD_WEATHER_ERROR,
    error,
  };
}
