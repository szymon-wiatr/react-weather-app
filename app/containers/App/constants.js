export const LOAD_WEATHER = 'weatherapp/App/LOAD_WEATHER';
export const LOAD_WEATHER_SUCCESS = 'weatherapp/App/LOAD_WEATHER_SUCCESS';
export const LOAD_WEATHER_ERROR = 'weatherapp/App/LOAD_WEATHER_ERROR';
