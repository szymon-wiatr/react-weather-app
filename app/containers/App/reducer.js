import produce from 'immer';
import {
  LOAD_WEATHER_SUCCESS,
  LOAD_WEATHER,
  LOAD_WEATHER_ERROR,
} from './constants';

export const initialState = {
  loading: false,
  error: false,
  currentCity: false,
  weatherData: {
    weather: false,
  },
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOAD_WEATHER:
        draft.loading = true;
        draft.error = false;
        draft.weatherData.weather = false;
        break;

      case LOAD_WEATHER_SUCCESS:
        [draft.weatherData.weather] = action.weather.data;
        draft.loading = false;
        draft.currentCity = action.cityname;
        break;

      case LOAD_WEATHER_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;
    }
  });

export default appReducer;
