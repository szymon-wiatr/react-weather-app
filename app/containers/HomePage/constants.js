export const CHANGE_CITYNAME = 'weatherapp/Home/CHANGE_CITYNAME';
