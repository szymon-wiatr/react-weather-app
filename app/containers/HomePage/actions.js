import { CHANGE_CITYNAME } from './constants';

/**
 * Changes the input field of the form
 *
 * @param  {string} cityname The new text of the input field
 *
 * @return {object} An action object with a type of CHANGE_CITYNAME
 */
export function changeCityname(cityname) {
  return {
    type: CHANGE_CITYNAME,
    cityname,
  };
}
