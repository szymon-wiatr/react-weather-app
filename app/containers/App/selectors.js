import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectWeather = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.weatherData.weather,
  );

export { selectGlobal, makeSelectLoading, makeSelectError, makeSelectWeather };
