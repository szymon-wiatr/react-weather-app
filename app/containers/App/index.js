import React from 'react';
import styled from 'styled-components';
import { Switch, Route } from 'react-router-dom';

import HomePage from 'containers/HomePage/Loadable';
import NotFoundPage from 'components/NotFoundPage/Loadable';

import GlobalStyle from '../../global-styles';

const AppWrapper = styled.div`
  max-width: calc(768px + 16px * 2);
  margin: 0 auto;
  padding: 16px;
`;

export default function App() {
  return (
    <AppWrapper>
      <Switch>
        <Route exact path="/" component={HomePage} />
        <Route path="" component={NotFoundPage} />
      </Switch>
      <GlobalStyle />
    </AppWrapper>
  );
}
