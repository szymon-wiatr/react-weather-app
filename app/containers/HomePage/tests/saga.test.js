import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_WEATHER } from 'containers/App/constants';
import { weatherLoaded, weatherLoadingError } from 'containers/App/actions';

import githubData, { getWeather } from '../saga';

const cityname = 'SkarÅ¼ysko-Kamienna';

/* eslint-disable redux-saga/yield-effects */
describe('getWeather Saga', () => {
  let getWeatherGenerator;

  beforeEach(() => {
    getWeatherGenerator = getWeather();

    const selectDescriptor = getWeatherGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getWeatherGenerator.next(cityname).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the weatherLoaded action if it requests the data successfully', () => {
    const response = [
      {
        name: 'Warsaw',
      },
      {
        name: 'London',
      },
    ];
    const putDescriptor = getWeatherGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(weatherLoaded(response, cityname)));
  });

  it('should call the weatherLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getWeatherGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(weatherLoadingError(response)));
  });
});

describe('githubDataSaga Saga', () => {
  const githubDataSaga = githubData();

  it('should start task to watch for LOAD_WEATHER action', () => {
    const takeLatestDescriptor = githubDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_WEATHER, getWeather));
  });
});
