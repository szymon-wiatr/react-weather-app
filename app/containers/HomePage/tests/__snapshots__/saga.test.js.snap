// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getWeather Saga should call the weatherLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getWeather Saga should call the weatherLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.weatherbit.io/v2.0/current?key=994a9b1c1fb546a2af4e7a7d92e118b2&city=Skarżysko-Kamienna",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getWeather Saga should dispatch the weatherLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getWeather Saga should dispatch the weatherLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.weatherbit.io/v2.0/current?key=994a9b1c1fb546a2af4e7a7d92e118b2&city=Skarżysko-Kamienna",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
